dev/core#4761 - CRM\Core\DAO - Generate unique name, case-insensitve

bsilvern opened the pull request:
Overview
----------------------------------------
Modified CRM\Core\DAO::makeNameFromLabel() to ensure that `name` values derived from the value of a label are unique in a case-insensitive context, as required by the `name` column's case-insensitive collation.

Before
----------------------------------------
* `name` was generated with case-sensitive uniqueness, causing failures when saving reminders or copying events. The modified function is believed to be used in other contexts as well, so there were potentially other failure modes.
* `name` was generated based on the value of label truncated at `columnWidth('name') - 3`. The 3-character suffix (an underscore + 2 digits) supported uniquifying only 100 names formed for any given value of label of that length or greater.

After
----------------------------------------
* `name` is generated with case-insensitive uniqueness, supporting both case-insensitive (default) and case-sensitive collation (should that ever be needed).
* A 5-character suffix is used (underscore + 4 alphanumerics), supporting generation of over 800,000 names from any given value of label (truncated at `columnWidth('name') - 5`).

Technical Details
----------------------------------------
underscore + 4 alphanumeric chars supports up to 36^4=1,679,616 unique names for any given value of label. However, as we approach all of the possible values for any given value of label being already present, the required number of guesses increases exponentially. With 50% of the possible values present, there is a 1/(2<sup>1</sup>) or 50% chance of failing to guess a unique value on the first try, and 1/(2<sup>20</sup>) or 0.000095% chance of failing to guess a unique value by the 20th try. It is expected that in all imaginable real scenarios, a unique value will be achieved on the first or second guess.

Comments
----------------------------------------
_Anything else you would like the reviewer to note_

Commit:
dev/core#4761 - CRM\Core\DAO - Generate unique name, case-insensitve

civicrm-builder commented:
Can one of the admins verify this patch?

civicrm-builder commented:
Can one of the admins verify this patch?

colemanw commented:
@civicrm-builder add to whitelist

colemanw commented:
Nice. @bsilvern you clearly have a computer-science background. Very strong first PR!
I like the way you've stepped up the limit for unique names, throwing letters into the mix increases the # of combinations quite a lot. The only thing we lose with that is the incremental-ness of the suffixes, but I don't think that's a real loss as every table in CiviCRM already has an `auto increment` id column.

Since you decided to think about very large numbers (`36^4` is pretty big I think)... I don't know if there's any real-world scenario for it, but ok I'll go there with you...

When you start to approach that scale of records, the resource limits you'll bump up against I think are a little different than what you described. Your analysis focused on the number of guesses that would be needed, but the guesses are made by a php loop of in-memory records, so the time-per-guess is negligible. However, fetching all those records form the database and loading them into memory in the first place is another matter. Anything close to the possible limit is likely to cause the whole script to run out of memory before it even gets a chance to make one guess, because right the query is set to retrieve every single match and store it in an array.

I can think of 2 ways to do this that would be faster and less likely to crash when working on that scale:

1. Keep the suffixes incremental, and use the `MAX()` sql function to fetch the 1 record with the largest suffix. You then know the value of your new suffix is that +1.
2. Keep the suffixes random. In the query, make 5 guesses (e.g. `WHERE `name` IN ("foo", "foo_6a51", "foo_0osa", "foo_c734", "foo_8ize")`. As long as the query returns <5 records, one of your guesses is available and you can use it. Otherwise, try again with 5 more guesses.

Thinking a bit more about it, I think 1 has some built-in-gotchas because it would be difficult to tell the difference between a suffixed duplicate vs a record with a longer name that happens to start with the same string as the original name. So if I had to write it I would chose method 2.

bsilvern commented:
@colemanw Thanks for your kind words. It's actually not my first PR, but it's been a while.

Agree with what you wrote about limitations of giant query results and your options # 1 and 2.

* Option 1: Yes, all the possibilities of what those names might look like give me pause to auto-increment. Especially true since the 5.66-alpha1 update code appended the id value to the name in civicrm_action_schedule. And there would be leading 0s (with and without) to consider. And after incrementing we'd have to test for overflow of the max width. Seems like a mess.

*  Option 2: Brilliant! But a bit more complicated (and perhaps bug-prone), and I'm concerned that other users are currently experiencing crashes when they mess with events and event reminders. So we must weigh getting the fix out ASAP vs. ability to support large numbers of similar names. I arrived here because my org has been using Civi for 12 years, and we had accumulated numerous events, most with reminders named "Reminder". Many events were copied from previous events, each time creating more "Reminder" names. But even then, we had only a few hundred such records. Is there a case where an unworkable number of matching names could be expected? I don't really have a good understanding of all the entities that have a name index, or even why the change was recently made to add indexes on the name column. If you think it's worthwhile, I'll rework this PR, but it may take me a couple of days (I have other responsibilities that need my attention).

I see some tests have failed. Mostly obvious things, like a test of the name field failing because it's been lower-cased. Some not so obvious, like "missing contribution", but I guess that could have the same cause. I chose to lower-case the stored name as I thought it was more intuitive given the default collation, but it was not necessary for correct operation. Should I update the PR to preserve the case of the name in order to pass the current tests, or should those failing tests be reworked? Modifying the tests is an area I'd rather not delve into. I currently have many failing tests on my civibuild Drupal9 installation (including ultimately a crash) even while testing the unmodified master branch. Not sure where I went wrong with the installation, but I didn't want to invest a lot of time investigating that.

colemanw commented:
@bsilvern sounds like we're mostly on the same page. I would vote to undo the lowercasing of names, at least for now to get this PR over the line.

The more I think about option 2 the more I like it. What potential bugs do you foresee?

Commit:
dev/core#4761 - CRM\Core\DAO - Generate unique name, case-insensitve

Replaces commit 9e4d050e5495bf9303d75b36c52f26fc8b3e2aa8

* Preserve case of name to prevent failure of existing
tests which expect the name to be unchanged when suffixes are not added to
create uniqueess.

* Implemented Coleman's suggestion to prevent memory blowout due to a
wildcard search with many matching names in the table.

bsilvern commented:
@colemanw `The more I think about option 2 the more I like it`. 
Yeah, me too. That's the right way to go. Added new commit. 
Thanks for the suggestion!

colemanw commented on CRM/Core/DAO.php:
```diff
@@ -3412,27 +3412,76 @@ private function makeNameFromLabel(): void {
+      // Make an initial attempt to guess a unique name by searching for
+      // 5 candidates (the original $name plus $name with 4 random suffixes).
+      // If all of these happen to exist in the table, we'll keep trying,
+      // doubling the number of guesses each time through the loop.
```
Clever. Doubling guesses with each iteration is like the inverse of a bisect, widening your net as you go out.

Commit:
dev/core#4761 - api\v4\SearchDisplay - mod test to accept random name suffix

SearchDisplayTest::testAutoFormatName was failing due to expecting _1 suffix
appended to name. Now will accept _r (r is any number of random chars).

eileenmcnaughton commented:
@colemanw @bsilvern where are we at here - this PR addresses a regression so ideally we would merge to the 5.68 branch before we cut the next rc

bsilvern commented:
I've implemented Coleman's suggestions and made a minor change to a test, so all tests pass now. As far as I'm concerned, it's ready to merge.

eileenmcnaughton commented:
thanks @bsilvern  - I don't support you are able to change which branch it is against or open a new one against the 5.68 branch? We would prefer to merge this to the rc to enable quicker release

bsilvern commented:
@eileenmcnaughton Well, I discovered that simply changing the base branch at the top of this page is NOT the way to go as that showed hundreds of changed files!

I'm sure there is some clever gitty way to change the branch of a PR, but I just manually recreated the changes on the 5.68 branch and submitted a new PR: https://github.com/civicrm/civicrm-core/pull/28352

eileenmcnaughton commented:
thanks @bsilvern  - yeah it can be tricky! I generally do something like a `git reset --hard origin/5.68 && git cherry-pick xyz`
